### PR TITLE
[rush] Create custom git hooks instead of copy

### LINKED ISSUE #3816: [rush] rush git-hooks edge case
## Summary

Background: our team is migrating husky to rush git-hooks, and we found an edge case with current rush git hooks implementation. So, this PR reports the edge case and proposes a alternative git hooks implementation.

## Repro steps

Assuming the repo maintainer wants to add a "rush prettier" to "pre-commit" git hooks. So the maintainer follows the official doc:
1. Add a `rush prettier` custom command
2. Use `rush prettier` in `common/git-hooks/pre-commit`.

Ok. Push and merge it to the main branch. Everything seems fine now.

Actually, users may run into an issue that the following issue throws:

```
rush: error: argument "<command>": Invalid choice: prettier (choose from ....
```

## Details

Let's say we got the following git branches with their commit history:

1. branch `main`: A -> B(rush prettier + pre-commit)
2. branch `feat-1`: B -> C (checkout from main when the latest commit is B)
3. branch `feat-2`: A -> D (checkout from main when the latest commit is A)

And B is the commit for `rush prettier` in pre-commit.

Now. Users checkout to `feat-1`, rush update to sync the git hooks to `.git` folder and correctly run `pre-commit` with `rush prettier`
And then, checkout to `feat-2`, modify a readme file, and run `git commit -m 'xxx'`, which triggers `pre-commit` hooks. BOOM, there is no `rush prettier` defined in branch `feat-2`.

### Possible solution

Instead of syncing files from `common/git-hooks` to `.git/hooks`, how about setting `core` by running `git config core.hooksPath common/git-hooks`

Similar logic in Husky at https://github.com/typicode/husky/blob/9efb720c165ffcfec774b4584aa61f59dae774c7/src/index.ts#L49

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? |  5.86.0 |
| `rushVersion` from rush.json? | 5.86.0 |
| `useWorkspaces` from rush.json? | Yes |
| Operating system? | Mac |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 14 |

### PR DESCRIPTION
**Pull request by stekycz**
## Summary

Fixes #3816

Resurrection of the work reverted in #3924 and originally implemented in #3873

## Details

I change copy of the script to create a new script from inlined template. The template might be another file but I chose simple solution now because the script is very simple. I can move it to a template file if needed but I would need guidance to where the file should be placed and what templating should be leveraged to hardcode specific values.

I am trying to keep the script universal to work even the directory of the repo is renamed. That is why I do not use absolute paths and I detect the script directory.

## How it was tested

I am not sure how to test these changes from the branch. However, I use the pattern of executing another script in one of my private repos.

## Impacted documentation

The [Git Hooks](https://rushjs.io/pages/maintainer/git_hooks/) documentation might need an update.

### CODE REVIEW

**Comment by stekycz** [comment]
Could anyone test it with Git LFS? I believe I have found the solution (see the inspiration in comment in the code) but I would like to test it on an existing repo before merging. Unfortunately, I do not have access to any LFS repo. Thank you!

**Comment by iclanton** [comment]
It'd be good to get a prerelease published that includes this change. @octogonz is more familiar with doing that than I am.

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts
Comment (stekycz) [reply libraries/rush-lib/src/logic/base/BaseInstallManager.ts:461]: I have no problem doing that, however, why is that better? Is it a convention in this repo? I usually assume using `path.join` can prevent possible issues with FS on various OS.

**Comment by stekycz** [comment]
I agree a prerelease would be nice to collect feedback and get it tested on a repo using LFS.

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts
Comment (dmichon-msft) [reply libraries/rush-lib/src/logic/base/BaseInstallManager.ts:461]: We've been trying to standardize on using `/` as the directory separator everywhere (the fs APIs all accept input in this format) because it simplifies path operations and makes them faster. It is still necessary to use `path.join` or `path.resolve` if one of the paths may contain `.` or `..` elements, though.

**Comment by iclanton** [comment]
@stekycz - This has been released as `@microsoft/rush@5.100.0-pr4132`.

**Comment by iclanton** [comment]
@stekycz - Looks like there's still a problem with LFS:

```
Handling LFS files
This should be run through Git's pre-push hook.  Run `git lfs update` to install it.
error: failed to push some refs to '<URL>'
```

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts
Comment (stekycz) [reply libraries/rush-lib/src/logic/base/BaseInstallManager.ts]: Updated

**Comment by stekycz** [comment]
@iclanton Got it, thanks for testing. I have made a few fixes. Is it possible to create a new pre-release pls?

On the way, I played a bit with LFS so I was able to test a few cases. I can do that once again with the pre-release version but it would be great if you could do it again 🙏 Thank you

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts
Comment (iclanton) [inline libraries/rush-lib/src/logic/base/BaseInstallManager.ts:471]:
```suggestion
          const gitLfsHookHandling: string = gitLfsHooks.has(filename)
```

Comment (stekycz) [reply libraries/rush-lib/src/logic/base/BaseInstallManager.ts]: Thank you! 🤦 🙏

**Comment by iclanton** [comment]
@stekycz - This has been released as @microsoft/rush@5.100.0-pr4132.1

**Comment by stekycz** [comment]
@iclanton Thanks. I have tried it using a few use cases I was able to come with. It seems working but I would prefer someone using LFS more to try it as well before merging 🙏

**Comment by iclanton** [comment]
It's working on a repo that's using LFS. Now I just want to try it on the same repo on a machine without LFS installed.

**Comment by iclanton** [comment]
There's a minor issue with line endings on Linux. The previous copy script normalized line endings. Now, I'm seeing an issue on this repo (rushstack) on WSL where the script under `common/git-hooks` has incorrect line endings. This manifests as an error message that looks like:

```
.git/hooks/pre-commit: /<path>/.git/hooks/../../common/git-hooks/pre-commit: /bin/sh^M: bad interpreter: No such file or directory
```

I'm not sure what can be done about that. I've tried forcing LF line endings on `/common/git-hooks/**/*` in `.gitattributes`, and that didn't work.

**Comment by stekycz** [comment]
Got it. I did the same thing as with permissions flags - change the file using correct line endings. However, there is still a change anyone would recall another script with wrong line endings. Let's consider that an edge case and responsibility of the repo maintainers. There is always a way how to ensure git hook scripts have correct line endings IMO.

**Review by iclanton — APPROVED** [review]

**Comment by ankhzet** [comment]
@iclanton I believe this breaks existing git-hooks that are relient on the arguments, passed to the hook:
```node common/scripts/install-run-rush.js --quiet commitlint --edit $1```
commitlint command here relies on the $1 arg passed by the git to the hook.
Solved by ammending `hooks/commit-msg` line 
```"$SCRIPT_IMPLEMENTATION_PATH"```
to
```"$SCRIPT_IMPLEMENTATION_PATH" $@```

**Comment by stekycz** [comment]
@ankhzet Good point! Thank you, I think it should be enough to modify this https://github.com/microsoft/rushstack/blob/main/libraries/rush-lib/src/logic/base/BaseInstallManager.ts#L490 from what you wrote, is that right?

**Comment by ankhzet** [comment]
@stekycz yes, seems the correct place. The exact edit is up to debate tho, `$@` should be ok to "forward args, passed to this script", but i'm unaware, if it would be universally supported on different target platforms.

**Comment by ankhzet** [comment]
@stekycz Another issue i've run into:
using explicit path to bash in the shebang line causes an issue with Github for Desktop on Windows (https://github.com/desktop/desktop/issues/11015):
If i change
```#!/bin/bash```
to 
```#!/bin/env sh```
it works in GithubDesktop without manually tweaking system's PATH env.

This issue is strictly relevant to a specific tooling and local environment configuration quirks tho, so can't say if it's critical. Althogh, it might be useful to include a warning in the changelog.

**Comment by stekycz** [comment]
@ankhzet Interesting! Would you, pls, send a PR to fix it? If it is just a one-liner, it should be easy to send a PR. Or, create an issue to track it at least 🙏 (instead of commenting on this PR). Btw, I would suggest to use `sh` instead of `bash` then as well. It may need a bit of tweaking of the script template in the code though.

**Pull request merged**
